### PR TITLE
4372 communities and groups 2.0

### PR DESCRIPTION
**Pull request by pkdash**
Implements https://github.com/hydroshare/hydroshare/issues/4372

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [ ] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. Test new features in Communities and Groups. https://app.moqups.com/90Oic9nPt1/view/page/ac8d4a7b1

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4747](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4747)

File: hs_communities/views/communities.py
Comment (devincowan) [inline hs_communities/views/communities.py:29]: is this necessary given L26?

File: hs_core/views/__init__.py
Comment (devincowan) [inline hs_core/views/__init__.py:1836]: this TODO seems pretty important--don't we want to implement that before we merge this PR?

File: hs_core/views/__init__.py
Comment (devincowan) [inline hs_core/views/__init__.py:2684]: remove dead code in this module

**Review by devincowan — CHANGES_REQUESTED** [review]
WOW! Impressive work.
Here is my first round of review.
This is focused mostly on the back end, didn't even look at the UI.
I plan a second review stage for the front end later this month.

File: hs_core/views/__init__.py
Comment (Maurier) [reply hs_core/views/__init__.py:1836]: This is no longer needed as an exception handler was added right below. I'll delete this comment.

File: config/hydroshare-config.yaml
Comment (pkdash) [reply config/hydroshare-config.yaml:6]: These local changes as part of running HS locally should have not been committed.

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py:39]: Good catch. We only need one of those events.

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py:115]: As part of handling the exception, I guess we can only log the error. Should we do that?

File: hs_access_control/management/scripts/bupusa.sh
Comment (pkdash) [reply hs_access_control/management/scripts/bupusa.sh:9]: I even don't know if we need this bash script anymore as now we have the UI to create any community. May be a question for @alvacouch

File: hs_access_control/management/scripts/resource.sh
Comment (pkdash) [reply hs_access_control/management/scripts/resource.sh:1]: I am wondering if we need this example script at all.

File: hs_access_control/models/community.py
Comment (pkdash) [reply hs_access_control/models/community.py:30]: Will check for community `closed` status before creating a join request record.

File: hs_communities/views/communities.py
Comment (Maurier) [reply hs_communities/views/communities.py]: It is not. I have removed the redundancy now.

File: hs_access_control/models/group.py
Comment (pkdash) [reply hs_access_control/models/group.py:85]: This is used in terms of creating or editing a group. The requirement says to send the email to the owners of the group. @horsburgh may want to chime in.

File: hs_access_control/models/user.py
Comment (pkdash) [reply hs_access_control/models/user.py:309]: I am guessing the use of the group email to allow any hydroshare user to contact the group via email. Again @horsburgh may chime in.

File: hs_access_control/models/user.py
Comment (pkdash) [reply hs_access_control/models/user.py:2825]: As  part of deleting a community request we delete the community that cascades to delete the associated request record. (https://github.com/hydroshare/hydroshare/blob/4372-communities-and-groups-2.0/hs_access_control/views.py#L853)

File: hs_access_control/management/commands/access_resource.py
Comment (pkdash) [reply hs_access_control/management/commands/access_resource.py:137]: Will fix it. Good catch.

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py:163]: Good suggestion.

File: hs_access_control/views.py
Comment (devincowan) [reply hs_access_control/views.py:258]: Good call.

File: hs_access_control/emails.py
Comment (devincowan) [reply hs_access_control/emails.py:115]: Might make sense to at least log it. I don't anticipate that it will happen, but doesn't hurt to catch and log

Comment (pkdash) [reply hs_access_control/emails.py:115]: Sounds good.

File: hs_access_control/models/invite.py
Comment (pkdash) [reply hs_access_control/models/invite.py:388]: It allows to approve a previously denied request to join a community. 
https://github.com/hydroshare/hydroshare/blob/4372-communities-and-groups-2.0/hs_access_control/views.py#L541

Comment (devincowan) [reply hs_access_control/models/invite.py:388]: Got it thanks

File: hs_access_control/models/user.py
Comment (pkdash) [reply hs_access_control/models/user.py:2870]: We have these assertions all over the access control app. I guess it serves two purposes:
 - to some extent validates that right parameter types are passed to methods/functions both during unit testing and manual testing locally.
 - it also documents parameter types expected for a function

File: config/hydroshare-config.yaml
Comment (pkdash) [reply config/hydroshare-config.yaml]: Fixed

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py:104]: Logging the exception.

File: hs_access_control/management/commands/access_resource.py
Comment (pkdash) [reply hs_access_control/management/commands/access_resource.py:137]: Fixed

File: hs_access_control/views.py
Comment (pkdash) [reply hs_access_control/views.py]: Refactored the GroupView and CommunityView classes.

File: hs_access_control/views.py
Comment (pkdash) [reply hs_access_control/views.py:88]: Looks like @Maurier has fixed it.

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py]: Fixed

File: hs_access_control/enums.py
Comment (pkdash) [reply hs_access_control/enums.py]: Fixed

File: hs_access_control/emails.py
Comment (pkdash) [reply hs_access_control/emails.py]: Done

File: hs_access_control/models/user.py
Comment (devincowan) [reply hs_access_control/models/user.py:2870]: Thanks @pkdash that of course makes sense for testing. I wasn't sure why we _only_ make the assertions when `__debug__=True` ...

File: hs_access_control/models/community.py
Comment (pkdash) [reply hs_access_control/models/community.py:30]: Now it checks for community `closed` status prior to creation of a request record for a group to join a community.

File: hs_access_control/models/group.py
Comment (horsburgh) [reply hs_access_control/models/group.py:85]: The group email address should probably be the email address of one of the group owners. It is possible that group owners might set up a separate email address for the group, but I'm guessing that won't be common. The primary purpose of the group email is for people (e.g., users who are not members of the group) to contact someone who is in charge of the group.

File: hs_access_control/models/user.py
Comment (horsburgh) [reply hs_access_control/models/user.py:309]: Yes - @pkdash has this right.

File: hs_access_control/models/group.py
Comment (Maurier) [reply hs_access_control/models/group.py:85]: @devincowan 

The group email defaults to the owner's email address in the Group creation form. 
image

It is displayed below the group description in the Group page.
image

File: hs_core/views/__init__.py
Comment (Maurier) [reply hs_core/views/__init__.py]: Removed

File: hsctl
Comment (pkdash) [reply hsctl:401]: line 403 has `rebuild_solr_index`.

File: hsctl
Comment (pkdash) [reply hsctl:427]: @alvacouch probably added this which is handy when using CUAHSI dev machines. These machines run out of space very quickly if the images are not deleted. Anyway, I will remove this change.

File: hs_access_control/management/scripts/resource.sh
Comment (pkdash) [reply hs_access_control/management/scripts/resource.sh:1]: @alvacouch Should we delete this folder (script) `hs_access_control/management/scripts`

File: hsctl
Comment (devincowan) [reply hsctl:401]: ha! Right you are thanks @pkdash

File: hs_access_control/models/user.py
Comment (pkdash) [reply hs_access_control/models/user.py:2870]: As I understand assertions are primarily used for the debugging purposes. I know we have some `assert` statements that run  in production.

Comment (devincowan) [reply hs_access_control/models/user.py:2870]: Fair enough, thanks @pkdash

File: hsctl
Comment (devincowan) [reply hsctl:427]: I guess if we only ever use `reset_all` on the dev rigs, then it is fine to have it include a `docker system prune -f`

Comment (pkdash) [reply hsctl:427]: Sounds good.

File: hs_access_control/management/scripts/bupusa.sh
Comment (pkdash) [reply hs_access_control/management/scripts/bupusa.sh:9]: @martinseul Do you see any need for having these scripts that @alvacouch created for managing communities from the command line when we did not had the UI to do that?  I checked with @Maurier on this as he was initially using these scripts. He suggests we should remove these scripts.

Comment (pkdash) [reply hs_access_control/management/scripts/bupusa.sh:9]: It seems @alvacouch wants to keep these scripts.

File: hs_communities/static/js/community.js
Comment (devincowan) [inline hs_communities/static/js/community.js:129]: do we need to do anything else to handle this?

File: hs_communities/static/js/community.js
Comment (devincowan) [inline hs_communities/static/js/community.js:150]: should we inform that the invite failed?

File: hs_communities/static/js/my-communities.js
Comment (devincowan) [inline hs_communities/static/js/my-communities.js:24]: more explanation than "Cancel Request"?

File: hs_communities/static/js/my-communities.js
Comment (devincowan) [inline hs_communities/static/js/my-communities.js:28]: notify user?

File: theme/static/js/hs-vue/group-communities-app.js
Comment (devincowan) [inline theme/static/js/hs-vue/group-communities-app.js:54]: message user?

File: theme/static/js/hs-vue/group-communities-app.js
Comment (devincowan) [inline theme/static/js/hs-vue/group-communities-app.js:79]: message user?

File: theme/static/js/hs-vue/group-communities-app.js
Comment (devincowan) [inline theme/static/js/hs-vue/group-communities-app.js:100]: message?

File: theme/static/js/hs-vue/group-communities-app.js
Comment (devincowan) [inline theme/static/js/hs-vue/group-communities-app.js:122]: message?

File: hs_communities/static/js/pending-community-request.js
Comment (devincowan) [inline hs_communities/static/js/pending-community-request.js:71]: message user if status is not 200?

File: hs_communities/static/js/pending-community-request.js
Comment (devincowan) [inline hs_communities/static/js/pending-community-request.js:86]:
```suggestion
```

File: hs_communities/static/js/pending-community-request.js
Comment (devincowan) [inline hs_communities/static/js/pending-community-request.js:148]: yes we should show error

**Review by devincowan — CHANGES_REQUESTED** [review]

**Comment by devincowan** [comment]
@Maurier I took a look at the front end work in this PR just now. It was a pleasure to review! :) I like the organization of your Vue apps and templates.

Most of my comments have to do with user messaging in the case of failed requests...

File: theme/static/css/hydroshare_core.css
Comment (Maurier) [reply theme/static/css/hydroshare_core.css:1569]: This was moved to another section below dedicated to flex helpers. https://github.com/hydroshare/hydroshare/blob/4372-communities-and-groups-2.0/theme/static/css/hydroshare_core.css#L3651

File: theme/static/js/hs-vue/group-members-app.js
Comment (Maurier) [reply theme/static/js/hs-vue/group-members-app.js:1]: I think I'll do this one as a future improvement. I've created an issue here: https://github.com/hydroshare/hydroshare/issues/4999

File: hs_communities/static/js/my-communities.js
Comment (Maurier) [reply hs_communities/static/js/my-communities.js:24]: I think the 'Cancel Request' title is consistent with our messaging where we use the name of the action being performed as the title, and the outcome of the action as the body. The outcome in this case is returned in the response object as `response.message`. I found the messages in this response user friendly enough.

https://github.com/hydroshare/hydroshare/blob/4372-communities-and-groups-2.0/hs_access_control/views.py#L823

File: hs_communities/static/js/community.js
Comment (Maurier) [reply hs_communities/static/js/community.js]: Added message.

File: hs_communities/static/js/community.js
Comment (Maurier) [reply hs_communities/static/js/community.js]: Added message.

File: theme/static/js/hs-vue/group-communities-app.js
Comment (Maurier) [reply theme/static/js/hs-vue/group-communities-app.js]: Added message.

File: theme/static/js/hs-vue/group-communities-app.js
Comment (Maurier) [reply theme/static/js/hs-vue/group-communities-app.js]: Added message.

File: hs_communities/static/js/pending-community-request.js
Comment (Maurier) [reply hs_communities/static/js/pending-community-request.js]: Added message.

File: hs_communities/static/js/pending-community-request.js
Comment (Maurier) [reply hs_communities/static/js/pending-community-request.js]: Removed

File: hs_communities/static/js/pending-community-request.js
Comment (Maurier) [reply hs_communities/static/js/pending-community-request.js]: Added message.

**Comment by Maurier** [comment]
@devincowan I went through these files and added messaging where they were missing.

File: theme/static/js/hs-vue/group-members-app.js
Comment (devincowan) [reply theme/static/js/hs-vue/group-members-app.js:1]: sounds good

File: hs_communities/static/js/my-communities.js
Comment (devincowan) [reply hs_communities/static/js/my-communities.js:24]: Good point, thanks

File: theme/static/css/hydroshare_core.css
Comment (devincowan) [reply theme/static/css/hydroshare_core.css:1569]: got it thank you

File: hs_communities/static/js/my-communities.js
Comment (Maurier) [reply hs_communities/static/js/my-communities.js]: Added message.

File: theme/static/js/hs-vue/group-communities-app.js
Comment (Maurier) [reply theme/static/js/hs-vue/group-communities-app.js]: Added message.

**Review by devincowan — DISMISSED** [review]
Awesome work @pkdash and @Maurier! Represents a pretty huge effort!
I hope we can get this on Beta soon...sorry for the delay
Will be nice to get final stakeholder approval and ship this thing!

**Comment by pkdash** [comment]
Thanks @devincowan for reviewing this large PR. You provided many valuable suggestions for improvements.

**Comment by pkdash** [comment]
@Maurier Please check my changes (last commit) to function group_json() where I added a parameter to include group owners. This function is used in other functions where you probably need to check and set this new parameter value accordingly.

File: hs_communities/views/communities.py
Comment (devincowan) [inline hs_communities/views/communities.py:83]: @pkdash I'm not seeing where we need this in the context but I bet you're right that we do...

Comment (pkdash) [reply hs_communities/views/communities.py:83]: If you try to add an owner, without this change the add owner modal won't be displayed. I guess it's not related to timeout issue though.

Comment (devincowan) [reply hs_communities/views/communities.py:83]: got it thanks

**Review by devincowan — APPROVED** [review]

**Review by Maurier — APPROVED** [review]

**Pull request merged**
